marimbabot_speech/src/text2command.py

```diff
@@ -1,48 +1,183 @@
+		"load test song",  # load the test song and play it
+	]
+	complicated_command_list = [
+		"play in 120 bpm",
+		"play in forte",
+		"play faster by 60 bpm",
+		"play slower by 40 bpm",
+		"play slower by 120 bpm",
+		"play louder by 20%",
+		"play softer by 40%",
+		# "preview",
+		# "preview slower by 30 bpm",
+		# "preview in 80 bpm",
+		# "preview in forte",
+	]
+	command_extractor = CommandExtraction()
+	command_list = simple_command_list + complicated_command_list
+	for command_text in command_list:
+		command_dict = command_extractor.extract(command_text)
+		print(f"{command_text}:{command_dict}")
+
+def dummy_clinet_command_subscriber():
+	# dummy client for command subscriber, for testing and instruction purpose
+	def command_callback(msg):
+		command_dict = json.loads(msg.command)
+		print(command_dict)
+	rospy.init_node('command_subscriber', log_level=rospy.DEBUG)
+	rospy.Subscriber('/speech_node/command', CommandMsg, command_callback, queue_size=10, tcp_nodelay=True)
+	rospy.spin()
 
-class CommandExtract:
+
+class CommandExtraction():
+	"""
+	Extract command from text, and publish it to the command topic.
+	"""
 	def __init__(self):
+		self.independent_action_list = [
+			"stop",
+			"start",
+			"repeat",
+			"loop",
+			"read",
+			"concatenate",
+			"save",
+			"load",
+		]
+		self.dependent_action_list = [
+			"play",
+			"preview",
+		]
+		self.speed_setup_list = [
+			"faster",
+			"slower",
+		]
+		self.volume_setup_list = [
+			"louder",
+			"softer",
+		]
 		self.command_pub = rospy.Publisher('/speech_node/command', CommandMsg, queue_size=100, tcp_nodelay=True)
 		self.speech_sub = rospy.Subscriber('/speech_node/speech', SpeechMsg, self.speech_callback, queue_size=10, tcp_nodelay=True)
-		self.command_list = get_commands()
-		self.hotword = "marimbabot"
-
-	def hotword_spotting(self, text):
-		if self.hotword in text:
-			rospy.logdebug(f"hotword spotted")
-			return True
-
-	def speech_callback(self, speech_msg):
-		if speech_msg.is_finished:
-			text = speech_msg.speech.lower()
-			rospy.logdebug(f"receive speech: {text}")
-			if self.hotword_spotting(text):
-				# if command in the same sentence of the spotting keyword, extract it
-				command = self.extract(text)
-				if command is not None:
-					self.pub_command(command)
-					rospy.logdebug(f"pub command: {command}")
-
-	def pub_command(self, command):
-		command_msg = CommandMsg()
-		command_msg.command = command
-		command_msg.header.stamp = rospy.Time.now()
-		self.command_pub.publish(command_msg)
-		rospy.logdebug(f"publish command: {command}")
-
-	def extract(self, text:str):
-		for command in self.command_list:
+
+	def speech_callback(self, req):
+		text = req.text
+		if req.no_speech_prob > 0.5:
+			rospy.logwarn("no_speech_prob > 0.5, but is passed the wad, something weird happened, ignore this speech")
+			return
+		command = self.extract(text)
+		rospy.logdebug(f"command extracted: {command}")
+		command_str = json.dumps(command)
```

Comment:
I think the JSON is much more flexible than the message, it's much easier to change when new requirements come in.

Reply:
And for interpreting the JSON, just one single line is enough.  `json.dumps(text)` and also for reuse of current msg.

And then, @scaredycode if you are okay with this usage, I will keep it, otherwise we need to talk about a new message format.